meta-balena-common/recipes-core/initrdscripts/files/migrate

```diff
@@ -102,12 +88,52 @@ migrate_enabled() {
     fi
     # Booting from external media - leave flashing after pivot-rooting
     umount "${FLASH_BOOT_MOUNT}" > /dev/null || true
-    relocate_log_mount
     return 1
 }
 
+# When copying to zram, we're unable to use traditional methods of limiting
+# memory usage such as ulimits and control groups. Memory used by zram is used
+# by the kernel. We can't just let the copy process cause an OOM, as we'd be
+# unable to effectively write anything to the logs for debugging afterwards.
+#
+# We also can't reliably predict how much memory we need to copy a given file
+# to zram, as the compression ratio varies at runtime depending on the file's
+# content.
+#
+# Instead, let's run the copy operation as a background job and keep an eye on
+# available memory. If it drops below the reserve we set, we'll kill the
+# operation and write a message to the log before bailing out.
+cp_memsafe() {
+    cp "${@}" &
+    pid=$!
+    available_reserve=$(( 16 * 1024 )) # 16 MB
+
+    while kill -0 $pid; do
+        mem_available=$(awk '/MemAvailable/{ print $2 }' /proc/meminfo)
+        if [ "${mem_available}" -lt "${available_reserve}" ]; then
+            kill $pid
+            fail "Unable to copy files, not enough memory"
+        fi
+    done
```

Comment:
I am instinctively taken back by complexity - I still think an early memory check, even if it won't be exact because of compression, is preferable to the added complexity of not having it.

Reply:
How do we do an early memory check and still allow for reduced memory usage due to the largely sparse nature and compressability of the images? I think just compressing the files and handling the case where we run out of memory is about as simple as it gets, though it's possible 16 MB of reserved memory isn't enough for the flasher.

Reply:
In testing, I've actually found (by altering the amount of memory available to QEMU) that the 16 MB reserve can allow the files to be copied to memory, but results in the OOM killer kicking in, and often taking out udevd.

We should probably increase this reserve to 24M or 32M at least, and possibly look at stopping udevd when running migrations.